vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -0,0 +1,666 @@
+#include "conditional_contextual_bandit.h"
+#include "reductions.h"
+#include "example.h"
+#include "global_data.h"
+#include "cache.h"
+#include "vw.h"
+#include "interactions.h"
+#include "label_dictionary.h"
+#include "cb_adf.h"
+#include "cb_algs.h"
+#include "constant.h"
+#include "v_array_pool.h"
+
+#include <numeric>
+#include <algorithm>
+#include <unordered_set>
+#include <queue>
+
+using namespace LEARNER;
+using namespace VW;
+using namespace VW::config;
+
+template <typename T>
+void return_v_array(v_array<T>& array, VW::v_array_pool<T>& pool)
+{
+  array.clear();
+  pool.return_object(array);
+}
+
+struct ccb
+{
+  vw* all;
+  example* shared;
+  std::vector<example*> actions, slots;
+  std::vector<uint32_t> origin_index;
+  CB::cb_class cb_label, default_cb_label;
+  std::vector<bool> exclude_list /*, include_list*/;
+  std::vector<std::string> generated_interactions;
+  std::vector<std::string>* original_interactions;
+  std::vector<CCB::label> stored_labels;
+  size_t action_with_label;
+
+  // All of these hashes are with a hasher seeded with the below namespace hash.
+  std::vector<uint64_t> hashes;
+  uint64_t id_namespace_hash;
+  std::string id_namespace_str;
+  
+  VW::v_array_pool<CB::cb_class> cb_label_pool;
+  VW::v_array_pool<ACTION_SCORE::action_score> action_score_pool;
+  VW::v_array_pool<ACTION_SCORE::action_scores> action_scores_pool;
+};
+
+namespace CCB
+{
+static constexpr uint32_t SHARED_EX_INDEX = 0;
+static constexpr uint32_t TOP_ACTION_INDEX = 0;
+
+void clear_all(ccb& data)
+{
+  data.shared = nullptr;
+  data.actions.clear();
+  data.slots.clear();
+  // data.exclude_list.clear();
+  // data.include_list.clear();
+  data.action_with_label = 0;
+}
+
+// split the slots, the actions and the shared example from the multiline example
+void split_multi_example(const multi_ex& examples, ccb& data)
+{
+  for (auto ex : examples)
+  {
+    switch (ex->l.conditional_contextual_bandit.type)
+    {
+      case example_type::shared:
+        data.shared = ex;
+        break;
+      case example_type::action:
+        data.actions.push_back(ex);
+        break;
+      case example_type::slot:
+        data.slots.push_back(ex);
+        break;
+      default:
+        THROW("ccb_adf_explore: badly formatted example - invalid example type");
+    }
```

Comment:
Return an error here to avoid throw.  Perhaps skip processing example?

Reply:
I don't believe we have a system to skip processing?

Reply:
What if we returned actions in the same order presented?

Reply:
This doesn't communicate the end user that this is a skipped iteration. It could be interpreted incorrectly easily.